### PR TITLE
CCO-318: Enable Azure Workload Identity authentication.

### PR DESCRIPTION
**Pull request by jstuever**
Authenticate with an Azure Workload Identity (AZWI) serviceaccount token when client secret is absent from auth config.

https://issues.redhat.com/browse/CCO-318

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -130,7 +130,7 @@ func New(config operatorconfig.Config, kubeConfig *rest.Config) (*Operator, erro
 		return nil, err
 	}
 	// example of future featuregate read and usage to set a variable to pass to a controller
-	_ = sets.NewString(enabled...).Has("AzureWorkloadIdentity")
+	AzureWorkloadIdentityEnabled := sets.NewString(enabled...).Has("AzureWorkloadIdentity")
```
Comment (JoelSpeed) [inline pkg/operator/operator.go:133]: You need to check if the feature is disabled, if it's disabled explicitly, then it shouldn't be enabled, even if it's in the enabled list IIUC

Comment (JoelSpeed) [reply pkg/operator/operator.go:133]: We discussed in slack, but, to close this off, lets get the PR updated with the latest library-go changes and this becomes very obvious, we are now only allowing to check if something is enabled, the mechanics internally mean you won't need to worry here

File: pkg/operator/operator.go [hunk 2]
```diff
@@ -128,6 +128,7 @@ func New(config operatorconfig.Config, kubeConfig *rest.Config) (*Operator, erro
 		return nil, err
 	}
 	// example of future featuregate read and usage to set a variable to pass to a controller
+	AzureWorkloadIdentityEnabled := featureGates.Enabled(configv1.FeatureGateAzureWorkloadIdentity)
```
Comment (Miciah) [inline pkg/operator/operator.go:131]: The comment can be removed (I missed that https://github.com/openshift/cluster-ingress-operator/pull/908/commits/3d26626ab5dd5d9fdcd8a4039892bbc4c173deef removed the `_ = sets.New[configv1.FeatureGateName](enabled...).Has("AzureWorkloadIdentity")` line that the comment described).  Uppercase `AzureWorkloadIdentityEnabled` looks funny for a function-local variable.  
```suggestion
	azureWorkloadIdentityEnabled := featureGates.Enabled(configv1.FeatureGateAzureWorkloadIdentity)
```

**Review by Miciah — COMMENTED** [review]
I've made some minor suggestions, but this looks fine to me overall.  
/approve

File: pkg/dns/azure/client/auth.go [hunk 3]
```diff
@@ -35,14 +37,65 @@ func getAuthorizerForResource(config Config) (autorest.Authorizer, error) {
...
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		options := azidentity.ClientSecretCredentialOptions{
+			ClientOptions: azcore.ClientOptions{
+				Cloud: cloudConfig,
+			},
+		}
+		cred, err = azidentity.NewClientSecretCredential(config.TenantID, config.ClientID, config.ClientSecret, &options)
```
[the remark this answers is not in the document]
Comment (Miciah) [reply pkg/dns/azure/client/auth.go]: Did you mean to delete `config.AzureWorkloadIdentityEnabled && ` in https://github.com/openshift/cluster-ingress-operator/compare/02a3c1434367c4ce82e84f3a42b393de92765541..d8c1cafad42eff9f781248a166e6e591a203ff24?

**Comment by Miciah** [comment]
Thanks!
/lgtm

**Comment by Miciah** [comment]
/hold
#939 is top priority; make sure it merges first.

**Comment by Miciah** [comment]
/hold cancel
and
/test all
now that #939 has merged.

**Comment by Miciah** [comment]
[e2e-gcp-operator  failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/906/pull-ci-openshift-cluster-ingress-operator-master-e2e-gcp-operator/1664461592286203904) because `TestInternalLoadBalancerGlobalAccessGCP` and `TestInternalLoadBalancer` failed.  This is a known issue ([OCPBUGS-13106](https://issues.redhat.com/browse/OCPBUGS-13106)), unrelated to the changes in this PR.
/override ci/prow/e2e-gcp-operator

**Pull request merged**